pkg/services/sqlstore/sqlstore_testinfra.go

```diff
@@ -305,6 +324,134 @@ func createTemporaryDatabase(tb TestingTB) (*testDB, error) {
+		// Each spannertest instance is a separate db.
+		srv, err := spannertest.NewServer("localhost:0")
+		if err != nil {
+			return nil, err
+		}
+		tb.Cleanup(srv.Close)
+
+		return &testDB{
+			Driver: "spanner",
+			Conn:   fmt.Sprintf("%s/projects/grafanatest/instances/grafanatest/databases/grafanatest;usePlainText=true", srv.Addr),
+		}, nil
+	}
+
+	conn := spannerDB
+	if spannerDB == "emulator" {
+		host := env("SPANNER_EMULATOR_HOST", "localhost:9010")
+		conn = fmt.Sprintf("%s/projects/grafanatest/instances/grafanatest/databases/grafanatest;usePlainText=true", host)
+	}
+
+	cfg, err := spannerdriver.ExtractConnectorConfig(conn)
+	if err != nil {
+		return nil, err
+	}
+
+	clientOptions := spannerConnectorConfigToClientOptions(cfg)
+
+	dbname := generateDatabaseName()
+	fullDbName := fmt.Sprintf("projects/%s/instances/%s/databases/%s", cfg.Project, cfg.Instance, dbname)
+	dbCreated := false
+
+	databaseAdminClient, err := database.NewDatabaseAdminClient(tb.Context(), clientOptions...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create database admin client: %v", err)
+	}
+	tb.Cleanup(func() {
+		if dbCreated {
+			// Drop database in the cleanup.
+			// Can't use tb.Context() here, since that is canceled before calling Cleanup functions.
+			err := databaseAdminClient.DropDatabase(context.Background(), &databasepb.DropDatabaseRequest{
+				Database: fullDbName,
+			})
+			if err != nil {
+				tb.Logf("Failed to drop Spanner database %s due to error %v", fullDbName, err)
+			} else {
+				tb.Logf("Dropped temporary Spanner database %s", fullDbName)
+			}
+		}
+
+		_ = databaseAdminClient.Close()
+	})
+
+	op, err := databaseAdminClient.CreateDatabase(tb.Context(), &databasepb.CreateDatabaseRequest{
+		Parent:          fmt.Sprintf("projects/%s/instances/%s", cfg.Project, cfg.Instance),
+		CreateStatement: fmt.Sprintf("CREATE DATABASE `%s`", dbname),
+		DatabaseDialect: databasepb.DatabaseDialect_GOOGLE_STANDARD_SQL,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create database: %v", err)
+	}
+	_, err = op.Wait(tb.Context())
+	if err != nil {
+		return nil, fmt.Errorf("failed to create database: %v", err)
+	}
+	tb.Logf("Created temporary Spanner database %s", fullDbName)
+
+	dbCreated = true
+
+	// Rebuild connection string, but change database to ID of just-created database.
+	// Example: `localhost:9010/projects/test-project/instances/test-instance/databases/test-database;usePlainText=true;disableRouteToLeader=true;enableEndToEndTracing=true`
+	buf := strings.Builder{}
+	if cfg.Host != "" {
+		buf.WriteString(cfg.Host)
+		buf.WriteString("/")
+	}
+	buf.WriteString("projects/")
+	buf.WriteString(cfg.Project)
+	buf.WriteString("/instances/")
+	buf.WriteString(cfg.Instance)
+	buf.WriteString("/databases/")
+	buf.WriteString(dbname) // Replace with new DB name.
+	for k, v := range cfg.Params {
+		buf.WriteString(";")
+		buf.WriteString(k)
+		buf.WriteString("=")
+		buf.WriteString(v)
+	}
```

Comment:
nit: This would be totally fine to do inefficiently with string concats & `fmt.Sprintf` for maintainability's sake as it's only used for tests.